Strict types validation (fix #48)

m0003r opened the pull request:
Resolves #48 

Validation is now strict, but query, path and cookie params (that are always strings) attempted to convert to specified type.

scaytrase commented:
Althougth the proposal itself looks valid I suggest to add more tests on HTTP related stuff. Because the problem I see the that, for example, path and query parameters are always strings by the nature of HTTP and how PHP process it. So considering any "number string" or "boolean string" to be string will break these things

scaytrase commented on src/Schema/Keywords/Type.php:
```diff
@@ -52,19 +49,17 @@ public function validate($data, string $type, ?string $format = null) : void
             case CebeType::BOOLEAN:
-                $stringifiedBool = is_scalar($data) && preg_match('#^(true|false)$#i', (string) $data);
-                if (! is_bool($data) && ! $stringifiedBool) {
+                if (is_string($data) || ! is_bool($data)) {
```
Looks redundant. If data is string it should be !is_bool as well, so right part is always true when left part is true and we can keep just right part

m0003r replied:
Fixed

scaytrase commented on src/Schema/Keywords/Type.php:
```diff
@@ -52,19 +49,17 @@ public function validate($data, string $type, ?string $format = null) : void
             case CebeType::INTEGER:
-                $stringifiedInt = is_scalar($data) && preg_match('#^[-+]?\d+$#', (string) $data) && ! is_float($data);
-                if (! is_int($data) && ! $stringifiedInt) {
+                if (is_string($data) || ! is_int($data)) {
```
same thoughts about redundancy here

scaytrase commented on src/PSR7/Validators/PathValidator.php:
```diff
@@ -49,8 +49,9 @@ private function validateRequest(OperationAddress $addr, RequestInterface $messa
         $validator        = new SchemaValidator($this->detectValidationStrategy($message));
 
         foreach ($pathParsedParams as $name => $value) {
+            $parameter = RequestParameter::fromSpec($specs[$name]);
```
I think with new usages this class should have new name. `SerializedParameter` maybe?

m0003r replied:
Renamed it

m0003r commented:
Thank you for quick reply

I have implemented deserialization for headers (including multipart POST) and cookies, but not yet for urlencoded post params. There would be significant amount of duplication between `QueryArgumentsValidator` and yet not existing `BodyArgumentsValidator`. Also there is similar code in `RequestCookieValidator` and `HeaderValidator`. May be a some new class should be introduced, taking array of specs and array of string values?

scaytrase commented:
>  There would be significant amount of duplication

Well, we have two ways, get working code with duplications to discuss refactoring or just skip this step if you have a plan already in your mind how it should look like

m0003r commented:
I have implemented deserialization also for POST parameters and added some tests.
If I understand correctly, `SerializedParameter` now can be also used to deserialize objects in [different styles](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#style-values), as mentioned in #2

scaytrase commented on src/PSR7/Validators/ArrayValidator.php:
```diff
@@ -0,0 +1,63 @@
+    {
+        foreach ($this->specs as $name => $spec) {
+            if ($spec->required && ! array_key_exists($name, $params)) {
+                throw RequiredParameterMissing::fromNameAndAddr($name, $addr);
```
Not really important but I generally prefer do a rethrow outside method with `$addr` instead of passing it to the method for only be used to construct an exception. This make further code readers think that some logic depends on `$addr` inside this method. This class look like low-level utility class and with moving `$addr` and PSR-7 specific exception could even live in `League\OpenAPIValidation\Schema` namespace

m0003r replied:
That's definitely a redundancy, I haven't noticed it

scaytrase commented on src/PSR7/Validators/CookiesValidator/ServerRequestCookieValidator.php:
```diff
@@ -37,38 +37,14 @@ public function __construct(array $specs)
+        try {
+            $validator->validateArray($addr, $message->getCookieParams(), $this->detectValidationStrategy($message));
+        } catch (RequiredParameterMissing $e) {
+            throw InvalidCookies::becauseOfMissingRequiredCookie($e->name(), $addr);
```
Actually, as far as I can see, you do the rethrow I mentioned here, so there is really no need to bound `ArrayValidtor` to PSR-7 stuff (`$addr`)

scaytrase commented on src/PSR7/Validators/PathValidator.php:
```diff
@@ -42,18 +41,14 @@ public function validate(OperationAddress $addr, MessageInterface $message) : vo
+            $validator->validateArray($addr, $pathParsedParams, $this->detectValidationStrategy($message));
+        } catch (InvalidParameter $e) {
+            throw InvalidPath::becauseValueDoesNotMatchSchema($e->name(), $e->value(), $addr, $e->getPrevious());
+        } // RequiredParameterMissing will not be thrown, all parameters are checking in parseParams
```
Well in this case I suggest rethrow to a `\LogicException`. This will never happen, but if unfortunatelly do - lib users could notify us to fix problem somewhere

scaytrase replied:
Without it any stat analyzer will think that this method will be able to throw `RequiredParameterMissing`

m0003r replied:
Good point, fixed it

scaytrase commented:
Actually I think I can fix it by myself

scaytrase commented:
Thanks @m0003r, sorry for long review

m0003r commented:
@scaytrase Thanks for the great library. I also have thoughts about XML validation (https://swagger.io/docs/specification/data-models/representing-xml/), maybe there will be another pull request sometime in the future :)

scaytrase commented:
Feel free to ping me if I forget to view it
